core/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodeFilters.java

```diff
@@ -64,21 +66,55 @@ public static DiscoveryNodeFilters buildFromKeyValue(OpType opType, Map<String,
         this.filters = filters;
     }
 
+    private Boolean match(String[] values, @Nullable String hostIp, @Nullable String publishIp) {
```

Comment:
can we make this mehtod just do a simple matching - returns true if any of the values matches any of hostIP and publishIP? then we don't need the Boolean and the null return. We then also push the crazy OpType dependent decision to the calling method where it's clearer why we do it.